src/mesh/forest/forest.cpp

```diff
@@ -0,0 +1,182 @@
+//========================================================================================
+// (C) (or copyright) 2024. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <algorithm>
+#include <array>
+#include <limits>
+#include <map>
+#include <memory>
+#include <set>
+#include <stack>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "defs.hpp"
+#include "mesh/forest/forest.hpp"
+#include "mesh/forest/relative_orientation.hpp"
+#include "mesh/forest/tree.hpp"
+#include "mesh/logical_location.hpp"
+#include "utils/bit_hacks.hpp"
+#include "utils/indexer.hpp"
+
+namespace parthenon {
+namespace forest {
+
+std::vector<LogicalLocation> Forest::GetMeshBlockListAndResolveGids() {
+  std::vector<LogicalLocation> mb_list;
+  std::uint64_t gid{0};
+  for (auto &[id, tree] : trees) {
+    std::size_t start = mb_list.size();
+    auto tree_mbs = tree->GetMeshBlockList();
+    mb_list.insert(mb_list.end(), std::make_move_iterator(tree_mbs.begin()),
+                   std::make_move_iterator(tree_mbs.end()));
+    std::size_t end = mb_list.size();
+    for (int i = start; i < end; ++i)
+      tree->InsertGid(mb_list[i], gid++);
+  }
+  // The index of blocks in this list corresponds to their gid
+  gids_resolved = true;
+  return mb_list;
+}
+
+Forest Forest::AthenaXX(RegionSize mesh_size, RegionSize block_size,
```

Comment:
That's the legacy Tree, isn't it?

Reply:
Yeah, this should build a forest that is equivalent to the single masked tree that Athena used for non-cubic base grids. (Equivalent in the sense that the block list retrieved for a grid built with this routine should be identical to that retrieved from the old tree code).

Reply:
I renamed this `Forest::HyperRectangular` for clarity.